### DISCUSSION TITLE
Central Pattern Generator Definition?

### DISCUSSION THREAD

**Discussion by aaprasad** [NeLy-EPFL/flygym#216] [Q&A] [answered]
Hi, I had a quick question about the CPG formulation. In Ijspeert 2007, they define the CPG system of oscillators as 
<img width="182" alt="image" src="https://github.com/user-attachments/assets/9a8dea77-8d85-4f6c-9938-6df09c71745f">

While in your paper you define it as: 
<img width="363" alt="image" src="https://github.com/user-attachments/assets/79cd82b8-beff-4453-a056-06d8688f937b">

I was wondering if you could give some intuition as to why you defined the differential equation governing $r$ as a first-order ODE rather than a second order and why you removed the $\frac{\alpha}{4}$ from your formulation. Thank you!

**Comment by sibocw** [comment, marked as answer]
Hello Aaditya,

Great question! This is a somewhat arbitrary simplification. It came from somewhere but I don't remember exactly where I saw it — likely it's from a course that's actually taught by Auke Ijspeert himself.

In any case, the intuition is simply that "if r is bigger than the nominal/target R, it should go down, otherwise it should go up, subject to some gain." The two dynamical systems are very similar. As an arbitrary example, if we choose R=1, a=10, r(0)=2, r'(0)=-1, then we can observe the following time evolution:
<img width="784" alt="image" src="https://github.com/user-attachments/assets/63e65572-5dc6-4262-bd54-de0d51eba95a">

Indeed, if we further plot the phase portrait and overlay the example trajectories on top, we observe:
<img width="774" alt="image" src="https://github.com/user-attachments/assets/caf8ccb2-8579-4f91-8830-00e3a98ed239">
so both dynamical systems have the same fixed point, and indeed both trajectories fall into it.

An important advantage of the original form from Ijspeert et al., 2007 is that it constraints the system so that |r'(t)| never gets too large (see how the blue line never goes below r'<2 in the first figure on the right). The intuition:
- By posing the y = gain(target - current state) form on the r (so that y is r'), the system in Wang-Chen et al ensures that r doesn't get too far off from its target (ie. R), but there's no constraint on r' itself.
- The original form from Ijspeert et al. goes one step further: by posing the same y = gain(target - current state) form on r' (so that y is r''), it also makes sure that r' never gets too far off from _its_ target.

Graphically, you can see that the arrows in the Quiver plot slowly push the state to the other side of the diagonal-ish line, where the r' gets pushed down toward 0.

A somewhat constrained |r'| might avoid numerical instabilities.

I've attached the plotting script below. Let me know if this answers your question.
[cpg_phase_portrait.py.zip](https://github.com/user-attachments/files/17077046/cpg_phase_portrait.py.zip)

**Reply by sibocw** [reply]
Actually, @stimpfli: maybe Nathan can try this out? Perhaps he can get away with a larger dt in his project.
